certora/specs/Splits.spec

```diff
@@ -0,0 +1,418 @@
+    uint128 splittableBefore; uint128 splittableAfter;
+
+    splittableBefore = splittable(e, receiver, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, receiver, assetId);
+
+    assert splittableAfter == splittableBefore + amt;
+}
+
+
+/// @notice The method give() gives funds from the user to the receiver.
+/// The splittable of any other user that is not receiver should not change
+rule splittableOfNonReceiverNotAffectedByGive() {
+    env e; uint256 userId; uint256 receiver; uint256 assetId; uint128 amt;
+    uint256 otherUser; uint128 splittableBefore; uint128 splittableAfter;
+
+    require otherUser != receiver;
+
+    splittableBefore = splittable(e, otherUser, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, otherUser, assetId);
+
+    assert splittableAfter == splittableBefore;
+}
+
+
+/// @notice the method hashSplits() Calculates the hash of the list of splits receivers.
+/// We use boolean selector to operate on two different lists of splits receivers.
+/// We verify that two calculated hashes are the same only if they got exactly the same input
+/// We also verify that different inputs must generate different hashes
+rule correctnessOfHashSplits() {
+    env e; uint256 index; uint256 length1; uint256 length2;  
+    uint256 userId1; uint32 weight1; uint256 userId2; uint32 weight2;
+    bytes32 receiversHash1; bytes32 receiversHash2;
+
+    length1 = getCurrSplitsReceiverLocaLength(e, true);
+    length2 = getCurrSplitsReceiverLocaLength(e, false);
+
+    userId1, weight1 = getCurrSplitsReceiverLocalArr(e, true, index);
+    userId2, weight2 = getCurrSplitsReceiverLocalArr(e, false, index);
+
+    receiversHash1 = hashSplits(e, true);
+    receiversHash2 = hashSplits(e, false);
+
+    assert (receiversHash1 == receiversHash2) => (length1 == length2);
+    assert (receiversHash1 == receiversHash2) => (userId1 == userId2);
+    assert (receiversHash1 == receiversHash2) => (weight1 == weight2);
+
+    assert (length1 != length2) => (receiversHash1 != receiversHash2);
+    assert (length1 == length2) && (userId1 != userId2) => (receiversHash1 != receiversHash2);
+    assert (length1 == length2) && (weight1 != weight2) => (receiversHash1 != receiversHash2);
+
+    // assert (receiversHash1 == receiversHash2) <=> ((length1 == length2) && (userId1 == userId2) && (weight1 == weight2));
+    // check with @Roy the commented line above
+}
+
+
+/// @notice Integrity of split:
+/// UserA has received drips and he has splittable > 0
+/// UserA has a list of splitters that should get some of the drips received
+/// UserA calls split()
+/// UserB is on the list of UserA's splitters, therefore "userB's splittable should NOT decrease"
+/// UserC is NOT on the list of UserA's splitters, therefore "userC's splittable should NOT change"
+/// UserA's collectable should NOT decrease, UserB and UserC's collectable should NOT change
+rule integrityOfSplit() {
```

Comment:
Can you improve this rule by checking the exact difference between the different users amount before and after, and not just checking that they decrease/increase?

Reply:
I think no, because I am testing for a general list of receivers (limited by loop iter to length of 2).
The exact amount that the split() returns is just how much was split to all the split receivers and the rest that is left as collectable to the splitter.
Therefore if there are more than one receivers, we don't know exactly how much each should get (unless we check each one's weight individually).

Reply:
Given that there are only two, can't you check the sum?

Reply:
Yes, I will add a forth user - userD who will be the second split receiver of userA and make sure that the sum of the increases in splittable of userB and userD is equal to the splittable of userA